### DISCUSSION TITLE
Parameters for Engine creation

### DISCUSSION THREAD

**Discussion by BStringhamVRSK** [google/filament#5470] [Ideas]
I would like to modify the `Engine::create()` to accept an optional struct containing parameters describing buffer sizes.  This would allow us to set memory usage needs at run-time, based on platform, scene size, etc., rather than recompiling the library.
**Question:**
Which buffer sizes make the most sense to expose?  The candidates I've considered are the ones set by these values:
- CONFIG_COMMAND_BUFFERS_SIZE
- CONFIG_PER_FRAME_COMMANDS_SIZE
- CONFIG_PER_RENDER_PASS_ARENA_SIZE

**Considerations:**
- What would be considered valid ranges for these?  
- I would expect to put the current values as the minimum (and default) values.
- Parameters would be passed in MB units

Thoughts?

EDIT:  Just found these in `Engine.h`.  Probably should include all of these as part of the configuration scheme?  Maybe I'll just do the memory ones for now.

```
    // TODO: these should come from a configuration object
    static constexpr float  CONFIG_Z_LIGHT_NEAR            = 5;
    static constexpr float  CONFIG_Z_LIGHT_FAR             = 100;
    static constexpr size_t CONFIG_FROXEL_SLICE_COUNT      = 16;
    static constexpr bool   CONFIG_IBL_USE_IRRADIANCE_MAP  = false;

    static constexpr size_t CONFIG_PER_RENDER_PASS_ARENA_SIZE   = filament::CONFIG_PER_RENDER_PASS_ARENA_SIZE;
    static constexpr size_t CONFIG_PER_FRAME_COMMANDS_SIZE      = filament::CONFIG_PER_FRAME_COMMANDS_SIZE;
    static constexpr size_t CONFIG_MIN_COMMAND_BUFFERS_SIZE     = filament::CONFIG_MIN_COMMAND_BUFFERS_SIZE;
    static constexpr size_t CONFIG_COMMAND_BUFFERS_SIZE         = filament::CONFIG_COMMAND_BUFFERS_SIZE;
````

**Comment by pixelflinger** [comment]
FYI this is something we've been wanting to do for a while -- it has never been a priority though. 

`CONFIG_FROXEL_SLICE_COUNT` should probably not be changed. 
`CONFIG_Z_LIGHT_NEAR` and `CONFIG_Z_LIGHT_FAR` should be handled like other options on View, I don't think these should be an engine parameter.

For the memory ones, I agree though. The important ones are:
`CONFIG_MIN_COMMAND_BUFFERS_SIZE` this one is the backend command queue. The needed size will depends mostly on how many objects you have in the scene. I would say that if you find you have to increase this on mobile, there is probably a problem with your scene, i.e. you have too many objects / draw calls. `CONFIG_COMMAND_BUFFERS_SIZE` should always be 3*`CONFIG_MIN_COMMAND_BUFFERS_SIZE` -- this is so we can have up to 3 frames of overlap. 


`CONFIG_PER_FRAME_COMMANDS_SIZE` and `CONFIG_PER_RENDER_PASS_ARENA_SIZE` are mostly used for the "high level" command buffer. Each command is 40 bytes I think. And again, its size depends on how many visible objects you have in the scene.

**Reply by romainguy** [reply]
It's tracked [here](https://github.com/google/filament/projects/10) and [here](https://github.com/google/filament/issues/2057).

**Reply by BStringhamVRSK** [reply]
Thanks for the feedback.  I am close to having a pull request put together, which will only touch the memory configuration.  I'll implement your recommendations.

The approach I'm taking adds an optional parameter to the `Engine::create()` call:
```
static Engine* create(Backend backend = Backend::DEFAULT,
        Platform* platform = nullptr, void* sharedGLContext = nullptr,
        const ConfigParams* configParams = nullptr);

```
If no configParams is provided, it will fall back to the default values indicated by FILAMENT_PER_RENDER_PASS_ARENA_SIZE_IN_MB, etc.

**Reply by romainguy** [reply]
The configuration should probably just be a struct passed by value? @pixelflinger would know what's best
